edb/lib/ext/ai.edgeql

```diff
@@ -0,0 +1,405 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2023-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+CREATE EXTENSION PACKAGE ai VERSION '1.0' {
+    set ext_module := "ext::ai";
+    set dependencies := ["pgvector==0.5"];
+
+    create module ext::ai;
+
+    create abstract type ext::ai::ProviderConfig extending cfg::ConfigObject {
+        create required property name: std::str {
+            set readonly := true;
+            create constraint exclusive;
+        };
+
+        create required property display_name: std::str {
+            set readonly := true;
+            create annotation std::description :=
+                "Human-friendly provider name";
+        };
+
+        create required property api_url: std::str {
+            set readonly := true;
+            create annotation std::description := "Provider API URL.";
+        };
+
+        create property client_id: std::str {
```

Comment:
Should this also be required? Without the client id, the server logs the error below when it has objects to index:
```
------------------------- Exception occurred: unhandled errors in a TaskGroup (1 sub-exception) --------------------------

1. ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)

------------------------------------------------------- Traceback --------------------------------------------------------

    /Users/james/dev/edgedb/edb/server/protocol/ai_ext.py, line 107, in _ext_ai_index_builder_controller_loop
        > processed = await _ext_ai_index_builder_work(
    /Users/james/dev/edgedb/edb/server/protocol/ai_ext.py, line 234, in _ext_ai_index_builder_work
        > async with asyncio.TaskGroup() as g:
    /opt/homebrew/Cellar/python@3.12/3.12.2_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/taskgroups.py, line 145, in __aexit__
        > raise me from None

ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
--------------------------------------------------- Grouped exceptions ---------------------------------------------------


    -------------------------- Exception occurred: 'NoneType' object has no attribute 'encode' ---------------------------

    1. AttributeError: 'NoneType' object has no attribute 'encode'

    ----------------------------------------------------- Traceback ------------------------------------------------------

        /Users/james/dev/edgedb/edb/server/protocol/ai_ext.py, line 329, in _generate_embeddings
            > return await _generate_openai_embeddings(provider, model_name, inputs)
        /Users/james/dev/edgedb/edb/server/protocol/ai_ext.py, line 341, in _generate_openai_embeddings
            > client = httpx.AsyncClient(
        /Users/james/dev/edgedb-dev/lib/python3.12/site-packages/httpx/_client.py, line 1371, in __init__
            > super().__init__(
        /Users/james/dev/edgedb-dev/lib/python3.12/site-packages/httpx/_client.py, line 182, in __init__
            > self.headers = Headers(headers)
        /Users/james/dev/edgedb-dev/lib/python3.12/site-packages/httpx/_models.py, line 75, in __init__
            > normalize_header_value(v, encoding),
        /Users/james/dev/edgedb-dev/lib/python3.12/site-packages/httpx/_utils.py, line 53, in normalize_header_value
            > return value.encode(encoding or "ascii")

    AttributeError: 'NoneType' object has no attribute 'encode'


ERROR 18430 _localdev 2024-04-12T14:11:40.496 edb.server.ai_ext: caught error in ext::ai::index builder on database 'main'
```
and requests to the rag endpoint return the same error:
```
curl --json '{
    "query": "What color is the sky on Mars?",
    "model": "gpt-4-turbo-preview",
    "context": {"query":"Astronomy"}
  }' http://localhost:5656/branch/main/ai/rag
AttributeError: 'NoneType' object has no attribute 'encode'%    
```

Reply:
Also when I removed the provider config to insert a new version with the client id, the server logged this error:
```
ERROR 21062 _localdev 2024-04-12T14:21:01.151 edb.server.ai_ext: ext::ai::index builder on database 'main': provider 'builtin::openai' has not been configured
```

Reply:
OpenAI doesn't require it, so I made it optional.